### ISSUE TITLE
Error with terragrunt version 0.67.0 and later when terragrunt-download-dir is on different drive than source code

### ISSUE THREAD

**Issue by vpham-houseloan** [gruntwork-io/terragrunt#3375]
## Describe the bug

Error when terragrunt-download-dir value on different drive than source code.

## Steps To Reproduce

Run a plan specifying ```--terragrunt-download-dir``` flag (or use ```TERRAGRUNT_DOWNLOAD``` environment variable) pointed to a different drive than the source code, e.g. with source code on ```D:\some_path\***\***\***```, and ```--terragrunt-download-dir``` value pointed to ```C:\tgdir```:
```
D:\some_path\***\***\***> terragrunt plan --terragrunt-download-dir C:\tgdir
```

If only one drive available, can be simulated on Windows with the following:
```
subst D: C:\
```

## Expected behavior

Able to plan without issues.

## Actual behavior

Error:
```
xx:xx:xx.xxx ERROR  Rel: can't make C:\tgdir\***\***\***\backend.tf relative to D:\some_path\***\***\***
```

## Workaround
Specify terragrunt-download-dir to be on the same drive or use a version earlier than 0.67.0.

## Versions

- Terragrunt version: 0.67.0 and later (0.67.1 at time of writing)
- Terraform version: 1.9.4
- Environment details (Ubuntu 20.04, Windows 10, etc.): Windows 10

## Additional context
Not sure if there's a corresponding issue on Linux or Mac.

**Comment by yhakbar** [comment]
Hey @vpham-houseloan , what's the earliest version you tested this on? 

I don't see anything in the release notes for [v0.67.0](https://github.com/gruntwork-io/terragrunt/releases/tag/v0.67.0), so I'm not sure what could have caused this bug. Are you saying that this error isn't encountered in [v0.66.9](https://github.com/gruntwork-io/terragrunt/releases/tag/v0.66.9)?

If you could provide a `terragrunt.hcl` file that could reproduce this issue and run your test with `--terragrunt-log-level debug`, that might help us understand what's going on here.

**Comment by vpham-houseloan** [comment]
> Hey @vpham-houseloan , what's the earliest version you tested this on?
> 
> I don't see anything in the release notes for [v0.67.0](https://github.com/gruntwork-io/terragrunt/releases/tag/v0.67.0), so I'm not sure what could have caused this bug. Are you saying that this error isn't encountered in [v0.66.9](https://github.com/gruntwork-io/terragrunt/releases/tag/v0.66.9)?

Correct. This error isn't encountered in v0.66.9 or earlier.

> 
> If you could provide a `terragrunt.hcl` file that could reproduce this issue and run your test with `--terragrunt-log-level debug`, that might help us understand what's going on here.


C:\some_path\a_path\sample\terragrunt.hcl:
```hcl
remote_state {
  config = {
    resource_group_name  = "***"
    storage_account_name = "***"
    container_name       = "***"
    key                  = "sample"
    use_azuread_auth     = true
  }
  backend = "azurerm"
  generate = {
    path      = "backend.tf"
    if_exists = "overwrite_terragrunt"
  }
}

terraform {
  source = "../sample-module"
}
```

C:\some_path\a_path\sample-module\main.tf:
```tf
terraform {
  required_providers {
    azurerm = {
      source  = "hashicorp/azurerm"
      version = "~> 3.0"
    }
  }
}

provider "azurerm" {
  features {}
}
```

with 
```powershell
subst D: C:\
cd D:\some_path\a_path\sample
```

v0.67.0:
```
D:\some_path\a_path\sample> terragrunt plan --terragrunt-log-level debug --terragrunt-download-dir C:\.tg
18:22:22.336 DEBUG  Terragrunt Version: 0.67.0
18:22:22.340 DEBUG  Did not find any locals block: skipping evaluation.
18:22:22.341 DEBUG  [.] Running command: terraform --version
18:22:22.525 DEBUG  terraform version: 1.9.4
18:22:22.525 DEBUG  Reading Terragrunt config file at terragrunt.hcl
18:22:22.529 DEBUG  Did not find any locals block: skipping evaluation.
18:22:22.530 DEBUG  Did not find any locals block: skipping evaluation.
18:22:22.531 INFO   Downloading Terraform configurations from file://D:/some_path/a_path/sample-module into C:/.tg/kIMKsvJByN4wwCDaxluYcc_qD_o/ulHLoMGzmvWx-MS40EKXUyKKgNI
18:22:22.536 DEBUG  Copying files from D:/some_path/a_path/sample into C:/.tg/kIMKsvJByN4wwCDaxluYcc_qD_o/ulHLoMGzmvWx-MS40EKXUyKKgNI
18:22:22.540 DEBUG  Setting working directory to C:/.tg/kIMKsvJByN4wwCDaxluYcc_qD_o/ulHLoMGzmvWx-MS40EKXUyKKgNI
18:22:22.542 ERROR  Rel: can't make C:\.tg\kIMKsvJByN4wwCDaxluYcc_qD_o\ulHLoMGzmvWx-MS40EKXUyKKgNI\backend.tf relative to D:\some_path\a_path\sample
18:22:22.542 ERROR  Unable to determine underlying exit code, so Terragrunt will exit with error code 1
```

v0.66.9:
```
D:\some_path\a_path\sample> terragrunt plan --terragrunt-log-level debug --terragrunt-download-dir C:\.tg
time=2024-09-03T18:20:45-05:00 level=debug msg=Terragrunt Version: 0.66.9
time=2024-09-03T18:20:45-05:00 level=debug msg=Did not find any locals block: skipping evaluation.
time=2024-09-03T18:20:45-05:00 level=debug msg=Running command: terraform --version prefix=[D:\some_path\a_path\sample]
time=2024-09-03T18:20:45-05:00 level=debug msg=terraform version: 1.9.4
time=2024-09-03T18:20:45-05:00 level=debug msg=Reading Terragrunt config file at D:\some_path\a_path\sample\terragrunt.hcl
time=2024-09-03T18:20:45-05:00 level=debug msg=Did not find any locals block: skipping evaluation.
time=2024-09-03T18:20:45-05:00 level=debug msg=Did not find any locals block: skipping evaluation.
time=2024-09-03T18:20:45-05:00 level=info msg=Downloading Terraform configurations from file://D:/some_path/a_path/sample-module into C:/.tg/kIMKsvJByN4wwCDaxluYcc_qD_o/ulHLoMGzmvWx-MS40EKXUyKKgNI
time=2024-09-03T18:20:45-05:00 level=debug msg=Copying files from D:/some_path/a_path/sample into C:/.tg/kIMKsvJByN4wwCDaxluYcc_qD_o/ulHLoMGzmvWx-MS40EKXUyKKgNI
time=2024-09-03T18:20:45-05:00 level=debug msg=Setting working directory to C:/.tg/kIMKsvJByN4wwCDaxluYcc_qD_o/ulHLoMGzmvWx-MS40EKXUyKKgNI
time=2024-09-03T18:20:45-05:00 level=debug msg=Generated file C:\.tg\kIMKsvJByN4wwCDaxluYcc_qD_o\ulHLoMGzmvWx-MS40EKXUyKKgNI\backend.tf. prefix=[D:\some_path\a_path\sample]
time=2024-09-03T18:20:45-05:00 level=debug msg=Initializing remote state for the azurerm backend prefix=[D:\some_path\a_path\sample]
time=2024-09-03T18:20:45-05:00 level=debug msg=Running command: terraform init prefix=[D:\some_path\a_path\sample]
Initializing the backend...

Successfully configured the backend "azurerm"! Terraform will automatically
use this backend unless the backend configuration changes.
Initializing provider plugins...
- Finding hashicorp/azurerm versions matching "~> 3.0"...
- Installing hashicorp/azurerm v3.116.0...
- Installed hashicorp/azurerm v3.116.0 (signed by HashiCorp)
Terraform has created a lock file .terraform.lock.hcl to record the provider
selections it made above. Include this file in your version control repository
so that Terraform can guarantee to make the same selections by default when
you run "terraform init" in the future.

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
time=2024-09-03T18:21:27-05:00 level=debug msg=Copying lock file from C:/.tg/kIMKsvJByN4wwCDaxluYcc_qD_o/ulHLoMGzmvWx-MS40EKXUyKKgNI/.terraform.lock.hcl to D:/some_path/a_path/sample prefix=[D:\some_path\a_path\sample]
time=2024-09-03T18:21:27-05:00 level=debug msg=Running command: terraform plan prefix=[D:\some_path\a_path\sample]

No changes. Your infrastructure matches the configuration.

Terraform has compared your real infrastructure against your configuration
and found no differences, so no changes are needed.
```

Cleanup:
```powershell
C:
subst D: /d
```

**Comment by vpham-houseloan** [comment]
[Possible culprit](https://github.com/gruntwork-io/terragrunt/compare/v0.66.9...v0.67.0#diff-25ca76d8ec427150ea1d2f9af53bbcc7fbab99ae1c2d6c8b567f1ff4c5fcdfaaL128-R137)

**Comment by vpham-houseloan** [comment]
Additional info:

Looks to affect more than the generate block. If I change ```if_exists``` to ```skip``` on the backend:

v0.67.0:
```
D:\some_path\a_path\sample> terragrunt plan --terragrunt-log-level debug --terragrunt-download-dir C:\.tg
19:29:19.506 DEBUG  Terragrunt Version: 0.67.0
19:29:19.510 DEBUG  Did not find any locals block: skipping evaluation.
19:29:19.511 DEBUG  [.] Running command: terraform --version
19:29:19.685 DEBUG  terraform version: 1.9.4
19:29:19.685 DEBUG  Reading Terragrunt config file at terragrunt.hcl
19:29:19.686 DEBUG  Did not find any locals block: skipping evaluation.
19:29:19.687 DEBUG  Did not find any locals block: skipping evaluation.
19:29:19.691 DEBUG  terraform files in C:/.tg/kIMKsvJByN4wwCDaxluYcc_qD_o/ulHLoMGzmvWx-MS40EKXUyKKgNI are up to date. Will not download again.
19:29:19.691 DEBUG  Copying files from D:/some_path/a_path/sample into C:/.tg/kIMKsvJByN4wwCDaxluYcc_qD_o/ulHLoMGzmvWx-MS40EKXUyKKgNI
19:29:19.696 DEBUG  Setting working directory to C:/.tg/kIMKsvJByN4wwCDaxluYcc_qD_o/ulHLoMGzmvWx-MS40EKXUyKKgNI
19:29:19.696 DEBUG  [.] The file path C:\.tg\kIMKsvJByN4wwCDaxluYcc_qD_o\ulHLoMGzmvWx-MS40EKXUyKKgNI\backend.tf already exists and if_exists for code generation set to "skip". Will not regenerate file.
19:29:19.697 DEBUG  [.] Initializing remote state for the azurerm backend
19:29:19.697 DEBUG  [.] Running command: terraform init
19:29:19.860 STDOUT [.] terraform: Initializing the backend...
19:29:24.931 STDOUT [.] terraform:
19:29:24.931 STDOUT [.] terraform: Successfully configured the backend "azurerm"! Terraform will automatically
19:29:24.931 STDOUT [.] terraform: use this backend unless the backend configuration changes.
19:29:25.271 STDOUT [.] terraform: Initializing provider plugins...
19:29:25.271 STDOUT [.] terraform: - Finding hashicorp/azurerm versions matching "~> 3.0"...
19:29:25.793 STDOUT [.] terraform: - Installing hashicorp/azurerm v3.116.0...
19:29:45.715 STDOUT [.] terraform: - Installed hashicorp/azurerm v3.116.0 (signed by HashiCorp)
19:29:45.716 STDOUT [.] terraform: Terraform has created a lock file .terraform.lock.hcl to record the provider
19:29:45.716 STDOUT [.] terraform: selections it made above. Include this file in your version control repository
19:29:45.716 STDOUT [.] terraform: so that Terraform can guarantee to make the same selections by default when
19:29:45.716 STDOUT [.] terraform: you run "terraform init" in the future.
19:29:45.720 STDOUT [.] terraform: Terraform has been successfully initialized!
19:29:45.720 STDOUT [.] terraform:
19:29:45.720 STDOUT [.] terraform: You may now begin working with Terraform. Try running "terraform plan" to see
19:29:45.720 STDOUT [.] terraform: any changes that are required for your infrastructure. All Terraform commands
19:29:45.720 STDOUT [.] terraform: should now work.
19:29:45.720 STDOUT [.] terraform: If you ever set or change modules or backend configuration for Terraform,
19:29:45.720 STDOUT [.] terraform: rerun this command to reinitialize your working directory. If you forget, other
19:29:45.720 STDOUT [.] terraform: commands will detect it and remind you to do so if necessary.
19:29:45.729 ERROR  1 error occurred:
        * Rel: can't make C:\.tg\kIMKsvJByN4wwCDaxluYcc_qD_o\ulHLoMGzmvWx-MS40EKXUyKKgNI\.terraform.lock.hcl relative to D:\some_path\a_path\sample


19:29:45.729 ERROR  Unable to determine underlying exit code, so Terragrunt will exit with error code 1
```

v0.66.9:
```
D:\some_path\a_path\sample> terragrunt plan --terragrunt-log-level debug --terragrunt-download-dir C:\.tg
time=2024-09-03T19:31:08-05:00 level=debug msg=Terragrunt Version: 0.66.9
time=2024-09-03T19:31:08-05:00 level=debug msg=Did not find any locals block: skipping evaluation.
time=2024-09-03T19:31:08-05:00 level=debug msg=Running command: terraform --version prefix=[D:\some_path\a_path\sample]
time=2024-09-03T19:31:09-05:00 level=debug msg=terraform version: 1.9.4
time=2024-09-03T19:31:09-05:00 level=debug msg=Reading Terragrunt config file at D:\some_path\a_path\sample\terragrunt.hcl
time=2024-09-03T19:31:09-05:00 level=debug msg=Did not find any locals block: skipping evaluation.
time=2024-09-03T19:31:09-05:00 level=debug msg=Did not find any locals block: skipping evaluation.
time=2024-09-03T19:31:09-05:00 level=debug msg=terraform files in C:/.tg/kIMKsvJByN4wwCDaxluYcc_qD_o/ulHLoMGzmvWx-MS40EKXUyKKgNI are up to date. Will not download again.
time=2024-09-03T19:31:09-05:00 level=debug msg=Copying files from D:/some_path/a_path/sample into C:/.tg/kIMKsvJByN4wwCDaxluYcc_qD_o/ulHLoMGzmvWx-MS40EKXUyKKgNI
time=2024-09-03T19:31:09-05:00 level=debug msg=Setting working directory to C:/.tg/kIMKsvJByN4wwCDaxluYcc_qD_o/ulHLoMGzmvWx-MS40EKXUyKKgNI
time=2024-09-03T19:31:09-05:00 level=debug msg=The file path C:\.tg\kIMKsvJByN4wwCDaxluYcc_qD_o\ulHLoMGzmvWx-MS40EKXUyKKgNI\backend.tf already exists and if_exists for code generation set to "skip". Will not regenerate file. prefix=[D:\some_path\a_path\sample]
time=2024-09-03T19:31:09-05:00 level=debug msg=Backend azurerm has not changed. prefix=[D:\some_path\a_path\sample]
time=2024-09-03T19:31:09-05:00 level=debug msg=Running command: terraform init prefix=[D:\some_path\a_path\sample]
Initializing the backend...
Initializing provider plugins...
- Reusing previous version of hashicorp/azurerm from the dependency lock file
- Using previously-installed hashicorp/azurerm v3.116.0

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.
time=2024-09-03T19:31:23-05:00 level=debug msg=Copying lock file from C:/.tg/kIMKsvJByN4wwCDaxluYcc_qD_o/ulHLoMGzmvWx-MS40EKXUyKKgNI/.terraform.lock.hcl to D:/some_path/a_path/sample prefix=[D:\some_path\a_path\sample]
time=2024-09-03T19:31:23-05:00 level=debug msg=Running command: terraform plan prefix=[D:\some_path\a_path\sample]

No changes. Your infrastructure matches the configuration.

Terraform has compared your real infrastructure against your configuration
and found no differences, so no changes are needed.
```

**Comment by vpham-houseloan** [comment]
Result is fine on 0.67.0 if --terragrunt-download-dir is on same drive (the following with ```if_exists``` set to ```overwrite_terragrunt```):

```
D:\some_path\a_path\sample> terragrunt plan --terragrunt-log-level debug --terragrunt-download-dir \.tg
20:33:30.003 DEBUG  Terragrunt Version: 0.67.0
20:33:30.007 DEBUG  Did not find any locals block: skipping evaluation.
20:33:30.008 DEBUG  [.] Running command: terraform --version
20:33:30.184 DEBUG  terraform version: 1.9.4
20:33:30.184 DEBUG  Reading Terragrunt config file at terragrunt.hcl
20:33:30.186 DEBUG  Did not find any locals block: skipping evaluation.
20:33:30.187 DEBUG  Did not find any locals block: skipping evaluation.
20:33:30.188 INFO   Downloading Terraform configurations from file://D:/some_path/a_path/sample-module into D:/.tg/kIMKsvJByN4wwCDaxluYcc_qD_o/ulHLoMGzmvWx-MS40EKXUyKKgNI
20:33:30.193 DEBUG  Copying files from D:/some_path/a_path/sample into D:/.tg/kIMKsvJByN4wwCDaxluYcc_qD_o/ulHLoMGzmvWx-MS40EKXUyKKgNI
20:33:30.195 DEBUG  Setting working directory to D:/.tg/kIMKsvJByN4wwCDaxluYcc_qD_o/ulHLoMGzmvWx-MS40EKXUyKKgNI
20:33:30.197 DEBUG  [.] Generated file ../../../.tg/kIMKsvJByN4wwCDaxluYcc_qD_o/ulHLoMGzmvWx-MS40EKXUyKKgNI/backend.tf.
20:33:30.198 DEBUG  [.] Initializing remote state for the azurerm backend
20:33:30.198 DEBUG  [.] Running command: terraform init
20:33:30.360 STDOUT [.] terraform: Initializing the backend...
20:33:36.746 STDOUT [.] terraform:
20:33:36.746 STDOUT [.] terraform: Successfully configured the backend "azurerm"! Terraform will automatically
20:33:36.746 STDOUT [.] terraform: use this backend unless the backend configuration changes.
20:33:37.295 STDOUT [.] terraform: Initializing provider plugins...
20:33:37.295 STDOUT [.] terraform: - Finding hashicorp/azurerm versions matching "~> 3.0"...
20:33:37.622 STDOUT [.] terraform: - Installing hashicorp/azurerm v3.116.0...
20:33:45.278 STDOUT [.] terraform: - Installed hashicorp/azurerm v3.116.0 (signed by HashiCorp)
20:33:45.278 STDOUT [.] terraform: Terraform has created a lock file .terraform.lock.hcl to record the provider
20:33:45.278 STDOUT [.] terraform: selections it made above. Include this file in your version control repository
20:33:45.278 STDOUT [.] terraform: so that Terraform can guarantee to make the same selections by default when
20:33:45.278 STDOUT [.] terraform: you run "terraform init" in the future.
20:33:45.285 STDOUT [.] terraform: Terraform has been successfully initialized!
20:33:45.285 STDOUT [.] terraform:
20:33:45.285 STDOUT [.] terraform: You may now begin working with Terraform. Try running "terraform plan" to see
20:33:45.285 STDOUT [.] terraform: any changes that are required for your infrastructure. All Terraform commands
20:33:45.285 STDOUT [.] terraform: should now work.
20:33:45.285 STDOUT [.] terraform: If you ever set or change modules or backend configuration for Terraform,
20:33:45.285 STDOUT [.] terraform: rerun this command to reinitialize your working directory. If you forget, other
20:33:45.285 STDOUT [.] terraform: commands will detect it and remind you to do so if necessary.
20:33:45.293 DEBUG  [.] Copying lock file from ../../../.tg/kIMKsvJByN4wwCDaxluYcc_qD_o/ulHLoMGzmvWx-MS40EKXUyKKgNI/.terraform.lock.hcl to .\
20:33:45.298 DEBUG  [.] Running command: terraform plan
20:33:52.443 STDOUT [.] terraform: Acquiring state lock. This may take a few moments...
20:33:55.219 STDOUT [.] terraform: No changes. Your infrastructure matches the configuration.
20:33:55.219 STDOUT [.] terraform:
20:33:55.220 STDOUT [.] terraform: Terraform has compared your real infrastructure against your configuration
20:33:55.220 STDOUT [.] terraform: and found no differences, so no changes are needed.
20:33:55.631 STDOUT [.] terraform: Releasing state lock. This may take a few moments...
```

Noting the following lines:
```
20:33:30.197 DEBUG  [.] Generated file ../../../.tg/kIMKsvJByN4wwCDaxluYcc_qD_o/ulHLoMGzmvWx-MS40EKXUyKKgNI/backend.tf.
```
```
20:33:45.293 DEBUG  [.] Copying lock file from ../../../.tg/kIMKsvJByN4wwCDaxluYcc_qD_o/ulHLoMGzmvWx-MS40EKXUyKKgNI/.terraform.lock.hcl to .\
```

**Comment by vpham-houseloan** [comment]
Resolved in [v0.67.5](https://github.com/gruntwork-io/terragrunt/releases/tag/v0.67.5) release.
